packages/core/strapi/src/node/vite/watch.ts

```diff
@@ -12,6 +13,33 @@ interface ViteWatcher {
 }
 
 const HMR_DEFAULT_PORT = 5173;
+const MAX_PORT_ATTEMPTS = 30;
+
+const findAvailablePort = (
```

Comment:
I kinda hate we have to do this. But if we're sure we don't have any solution then why not

Out of curiosity, wouldn't that create less noise and consume less resources to use net.Socket and try establishing connections rather than create actual servers?

Reply:
Yeah, I hate it too. If it affected production I wouldn't do it, but since it's just dev mode I think it's acceptable.

This create server method seemed the most reliable since it actually binds to a port it's certain it's free instead of just being somewhat sure. And slightly less complex because trying to connect to a port requires more listeners. But yeah, it is slightly more "noisy" as you said.

Reply:
It would be nice to start unit testing all this stuff with server mocks etc... even though for now we don't have any :thinking:
This is the kind of behavior that I don't want anyone to break because they don't have context or smth :confused:

Reply:
Next week is a testing blitz... maybe you can come up with something 😆